server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3502,11 +3709,42 @@ SnapshotsInProgress.Entry computeUpdatedSnapshotEntryFromShardUpdates() {
                             + clonesBuilder
                             + " as well as "
                             + shardsBuilder;
+
+                    // Check horizontally within the snapshot to see whether any previously limited shard snapshots can now start
+                    maybeStartAssignedQueuedShardSnapshots(
```

Comment:
This has a side effect that a snapshot will kickoff assigned-queued shards belong to the same snapshot before starting shard snapshot of the same shard but belongs to a later snapshot. I think this is desirable. It means once a shard finishes its snapshot, it does _not_ immediately get another snapshot in `INIT` state, i.e. something we previously wanted to prevent with limiting the number of concurrent snapshots. With limiting number of `INIT` shards, it is less important now but still nice to have.